File: BackPropagationCS/NetworkAlgorithm.cs
using NeuronNetwork.Structure;
using static System.Math;

namespace NeuronNetwork.Algorithm
{
    public class Algorithms
    {
        public enum NormalizationMode
        {
            ZeroToOne = 0,
            MinusOneToOne = 1
        };

        public static double Sigmoid(double x, double response)
        {
            return 1.0 / (1.0 + Exp(-x * response));
        }

        public static double Sigmoid_D(double x)
        {
            return x * (1.0 - x);
        }

        public static double ShiftedSigmoid(double x, double response)
        {
            return Sigmoid(x, response) * 2.0 - 1.0;
        }

        public static double ShiftedSigmoid_D(double x)
        {
            return Sigmoid_D((x + 1.0) / 2.0);
        }

        public static double ReLU(double x, double response)
        {
            double _x = x * response;
            return _x > 0.0 ? _x : 0.0;
        }

        public static double ReLU_D(double x)
        {
            return x > 0.0 ? 1.0 : 0.0;
        }

        public static double[] Normalization(byte[] data, NormalizationMode mode)
        {
            double[] o = new double[data.Length];

            if (mode == NormalizationMode.ZeroToOne)
            {
                for (int i = 0; i < o.Length; i++)
                {
                    o[i] = data[i] / 255.0;
                }
            }
            else
            {
                for (int i = 0; i < o.Length; i++)
                {
                    o[i] = data[i] / 127.5 - 1.0;
                }
            }

            return o;
        }

        public static int GetIntFromBytes(byte[] data, int offset, bool MSB = false)
        {
            byte[] _data = ArrayExtension.Crop(data, offset, sizeof(int));

            if (MSB) Array.Reverse(_data);

            return BitConverter.ToInt32(_data, 0);
        }

        public static void RandomDist(NeuronLayer layer, double min = 0.0, double max = 1.0)
        {
            Random rnd = new Random(Environment.TickCount);
            for (int i = 0; i < layer.neurons.Length; i++)
            {
                for (int j = 0; j < layer.neurons[i].weights.Length; j++)
                {
                    layer.neurons[i].weights[j] = CastToRange(rnd.NextDouble(), min, max);
                }
            }
        }

        public static double CastToRange(double d, double min, double max)
        {
            return min + d * (max - min);
        }
    }
}

namespace System
{
    public class ArrayExtension
    {
        public static T[] Crop<T>(T[] array, int start, int length)
        {
            return array.Skip(start).Take(length).ToArray();
        }
    }
}

namespace NeuronNetwork.Structure
{
    public partial class NeuronLayer
    {
        public void SoftMax()
        {
            double sum = 0.0;
            double biggestValue = neurons[0].value;

            for (int i = 0; i < neurons.Length; i++)
            {
                if (neurons[i].value > biggestValue)
                    biggestValue = neurons[i].value;
            }

            for (int i = 0; i < neurons.Length; i++)
            {
                sum += Exp(neurons[i].value - biggestValue);
            }

            for (int i = 0; i < neurons.Length; i++)
            {
                neurons[i].value = Exp(neurons[i].value - biggestValue) / sum;
            }
        }

        public void SoftMax_GetError(double[] target)
        {
            if(target.Length != neurons.Length)
            {
                throw new Exception("Target datasize mismatch!");
            }

            for(int i = 0; i < neurons.Length; i++)
            {
                neurons[i].error = target[i] - neurons[i].value;
            }
        }
    }
}

File: NeuronNetworkCpp/NetworkAlgorithm.cpp
#include "NetworkAlgorithm.h"

#include<math.h>

using namespace Network::Algorithm;
typedef Network::float_n float_n;

float_n Network::Algorithm::Sigmoid(float_n x)
{
	return 1.0 / (1.0 + exp(-x));
}

float_n Network::Algorithm::Sigmoid_D(float_n x)
{
	return x * (1.0 - x);
}

float_n Network::Algorithm::ShiftedSigmoid(float_n x)
{
	return Sigmoid(x) * 2.0 - 1.0;
}

float_n Network::Algorithm::ShiftedSigmoid_D(float_n x)
{
	return Sigmoid_D((x + 1.0) / 2.0);
}

float_n Network::Algorithm::ReLU(float_n x)
{
	return x > 0.0 ? x : 0.0;
}

float_n Network::Algorithm::ReLU_D(float_n x)
{
	return x > 0.0 ? 1.0 : 0.0;
}

#define LEAKY_RELU_CONST 0.01

float_n Network::Algorithm::LeakyReLU(float_n x)
{
	return x > 0.0 ? x : x * LEAKY_RELU_CONST;
}

float_n Network::Algorithm::LeakyReLU_D(float_n x)
{
	return x > 0.0 ? 1.0 : LEAKY_RELU_CONST;
}

// NOTE: Added offset (2023-2-20)
void Normalization_ZeroToOne(float* dataOut, int offset, unsigned char* data, int dataSize)
{
	for (int i = 0; i < dataSize; i++)
	{
		dataOut[i] = data[i + offset] / 256.0;
	}
}

void Normalization_MinusOneToOne(float* dataOut, int offset, unsigned char* data, int dataSize)
{
	for (int i = 0; i < dataSize; i++)
	{
		dataOut[i] = data[i + offset] / 128.0 - 1.0;
	}
}

void Normalization_NoNormalization(float* dataOut, int offset, unsigned char* data, int dataSize)
{
	for (int i = 0; i < dataSize; i++)
	{
		dataOut[i] = data[i + offset];
	}
}

float* Network::Algorithm::NormalizeData(unsigned char* data, int offset, int dataSize, NormalizationMode mode)
{
	float* dataOut = new float[dataSize];

	switch (mode)
	{
	case NormalizationMode::ZeroToOne:
		Normalization_ZeroToOne(dataOut, offset, data, dataSize);
		break;

	case NormalizationMode::MinusOneToOne:
		Normalization_MinusOneToOne(dataOut, offset, data, dataSize);
		break;

	default:
		Normalization_NoNormalization(dataOut, offset, data, dataSize);
		break;
	}

	return dataOut;
}

void Network::Algorithm::SoftMax(Network::NeuronLayer& layer)
{
	double sum = 0.0;
	double biggestValue = layer[0].value;

	// find largest value in neurons
	for (int i = 0; i < layer.Count(); i++)
	{
		if (layer[i].value > biggestValue)
			biggestValue = layer[i].value;
	}

	// add up sums
	for (int i = 0; i < layer.Count(); i++)
	{
		sum += exp(layer[i].value - biggestValue);
	}

	// set values
	for (int i = 0; i < layer.Count(); i++)
	{
		layer[i].value = exp(layer[i].value - biggestValue) / sum;
	}
}

void Network::Algorithm::SoftMax(Network::NeuronLayerInstance& layer)
{
	double sum = 0.0;
	double biggestValue = layer[0].value;

	// find largest value in neurons
	for (int i = 0; i < layer.Count(); i++)
	{
		if (layer[i].value > biggestValue)
			biggestValue = layer[i].value;
	}

	// add up sums
	for (int i = 0; i < layer.Count(); i++)
	{
		sum += exp(layer[i].value - biggestValue);
	}

	// set values
	for (int i = 0; i < layer.Count(); i++)
	{
		layer[i].value = exp(layer[i].value - biggestValue) / sum;
	}
}

void Network::Algorithm::SoftMaxGetError(NeuronLayer& layer, float_n* target)
{
	for (int i = 0; i < layer.Count(); i++)
	{
		layer[i].error = target[i] - layer[i].value;
	}
}

void Network::Algorithm::SoftMaxGetError(NeuronLayerInstance& layer, float_n* target)
{
	for (int i = 0; i < layer.Count(); i++)
	{
		layer[i].error = target[i] - layer[i].value;
	}
}